pkg/streamingpromql/operators/aggregations/limitklimitratio/utils.go

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package limitklimitratio
+
+import (
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+type quietCloser interface {
+	close()
+}
+
+type stepCounter interface {
+	quietCloser
+	inc(step int)
+	count(step int) int
+}
+
+type stepCounterNoOp struct{}
+
+func (s *stepCounterNoOp) inc(_ int)       {}
+func (s *stepCounterNoOp) close()          {}
+func (s *stepCounterNoOp) count(_ int) int { return 0 }
+
+type stepCounterImpl struct {
+	// index by step --> number of series (which have a point at this step) we have included for the group
+	seriesIncludedAtStep     []int
+	memoryConsumptionTracker *limiter.MemoryConsumptionTracker
+}
+
+func (s *stepCounterImpl) close() {
+	types.IntSlicePool.Put(&s.seriesIncludedAtStep, s.memoryConsumptionTracker)
+}
+
+func (s *stepCounterImpl) inc(step int) {
+	s.seriesIncludedAtStep[step]++
+}
+
+func (s *stepCounterImpl) count(step int) int {
+	return s.seriesIncludedAtStep[step]
+}
+
+func newStepCounterImpl(size int, memoryConsumptionTracker *limiter.MemoryConsumptionTracker) (stepCounter, error) {
+	intSlice, err := types.IntSlicePool.Get(size, memoryConsumptionTracker)
+	if err != nil {
+		return nil, err
+	}
+	intSlice = intSlice[:size]
+	return &stepCounterImpl{
+		memoryConsumptionTracker: memoryConsumptionTracker,
+		seriesIncludedAtStep:     intSlice,
+	}, nil
+}
+
+type limitGroup struct {
+	seriesCount              int            // Total number of series that contribute to this group.
+	filled                   bool           // Flag to indicate we have identified k series for this group and no more series need to be accumulated
+	accumulatedSeries        []*querySeries // A place to store the series we will return for this group
+	stepCounter              stepCounter    // A utility to track the number of series points for each step across the group
+	memoryConsumptionTracker *limiter.MemoryConsumptionTracker
+}
+
+func (q *limitGroup) initAccumulatedSeries(size int) {
```

Comment:
[nit] Perhaps `seriesCount` would be a clearer name rather than `size`? (It took me a second to understand if this was the number of series or number of steps, or something else.)